Tools/clinic/clinic.py

```diff
@@ -4276,11 +4282,11 @@ def eval_ast_expr(
 
 
 class IndentStack:
-    def __init__(self):
-        self.indents = []
-        self.margin = None
+    def __init__(self) -> None:
+        self.indents: list[int] = []
+        self.margin = ""
```

Comment:
The way this class is written currently is a bit of a code smell (my instinct is that `margin` probably shouldn't be an instance attribute at all), but I feel like this change might cover up the problem rather than fix it. I think you shouldn't really be calling `.indent()` or `.dedent()` on this class before calling `.infer()` to set the margin, so maybe we should do something like this? (Diff is relative to your PR branch):

```diff
--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -4284,7 +4284,7 @@ def eval_ast_expr(
 class IndentStack:
     def __init__(self) -> None:
         self.indents: list[int] = []
-        self.margin = ""
+        self.margin: str | None = None

     def _ensure(self) -> None:
         if not self.indents:
@@ -4347,6 +4347,7 @@ def indent(self, line: str) -> str:
         """
         Indents a line by the currently defined margin.
         """
+        assert self.margin is not None, "Cannot call .indent() before calling .infer()"
         return self.margin + line

     def dedent(self, line: str) -> str:
@@ -4354,6 +4355,7 @@ def dedent(self, line: str) -> str:
         Dedents a line by the currently defined margin.
         (The inverse of 'indent'.)
         """
+        assert self.margin is not None, "Cannot call .indent() before calling .infer()"
         margin = self.margin
         indent = self.indents[-1]
         if not line.startswith(margin):
@@ -5238,6 +5240,7 @@ def parse_slash(self, function: Function) -> None:
             p.kind = inspect.Parameter.POSITIONAL_ONLY

     def state_parameter_docstring_start(self, line: str | None) -> None:
+        assert self.indent.margin is not None, "self.margin.infer() has not yet been called to set the margin"
         self.parameter_docstring_indent = len(self.indent.margin)
         assert self.indent.depth == 3
         return self.next(self.state_parameter_docstring, line)
```

Reply:
(The `assert`s are ugly and verbose, but only because they reveal the underlying code smell -- at least they're honest ;)

Reply:
Yeah, I thought about this, but went for the easy version ;) Using the asserts, it's easier to identify the smells, as you say.

Reply:
Applied!